smac/utils/configspace.py

```diff
@@ -169,12 +169,16 @@ def print_config_changes(
     if incumbent is None or challenger is None:
         return
 
-    params = sorted([(param, incumbent[param], challenger[param]) for param in challenger.keys()])
-    for param in params:
-        if param[1] != param[2]:
-            logger.info("--- %s: %r -> %r" % param)
-        else:
-            logger.debug("--- %s Remains unchanged: %r", param[0], param[1])
+    all_keys = set(incumbent.keys().union(challenger.keys()))
+
+    params = sorted(
+        [(param, incumbent.get(param), challenger.get(param)) for param in all_keys]
+    )
+    for (key, inc_val, challenger_val) in params:
+        msg = f"--- {key}: {inc_val} -> {challenger_val}"
+        if inc_val == challenger_val:
+            msg += " (unchanged)"
+        logger.debug(msg)
```

Comment:
In reference to #1041, I propose simplifying the generation of debug messages and the procedure of the entire loop. The current implementation generates a list of parameter tuples using a list comprehension (essentially a loop) and then iterates over this list using a distinct for-in loop. Instead, these operations can be consolidated into a single loop that explicitly iterates over the sorted keys:

```python
    for key in sorted(all_keys):
        incumbent_val = incumbent.get(key)
        challenger_val = challenger.get(key)
        logger.debug(f"--- {key}: {incumbent_val} -> {challenger_val}"
                     f"{' (unchanged)' if incumbent_val == challenger_val else ''}")
```

This change simplifies the code by removing the need for the intermediate `params` list (i.e, variable, and list comprehension), and potentially improves efficiency by reducing the number of loops (yet they were not nested so not a big deal anyway). Nonetheless, it also enhances readability by generating the complete debug message string within the logger.debug() call, using a ternary expression to optionally append ' (unchanged)' rather than concatenation.

Your thoughts?

Reply:
Yup this should be faster, good suggestion :) I was just trying to not deviate too much from the original code.

I think in practice, it would be faster to build up the entire string and debug message at once, rather than logging each line. Given the `logger.debug` requires a syscall, doing as little of these as possible in a parallel set up makes sense